### PR TITLE
Collection ops - contains; in; empty literal

### PR DESCRIPTION
**Pull request by niquola**

### CODE REVIEW

File: src/equality.js [hunk 1]
```diff
@@ -69,5 +69,26 @@ engine.gte = function(a, b){
   return a[0] >= b[0];
 };
 
+function containsImpl(a,b){
```
Comment (plynchnlm) [inline src/equality.js]: This could use a function comment explaining that b is assumed to have one element, and that it tests whether b[0] is in a.

**Comment by plynchnlm** [comment]
In the README, to the list of implemented sections you can now add "6.4 Collections".  Also, it looks like "6.2 Comparisons" is done, and "6.5. Boolean logic".

**Comment by plynchnlm** [comment]
Would you like to set the version to 0.7.0 and update CHANGELOG.md?

File: src/equality.js [hunk 2]
```diff
@@ -69,5 +69,26 @@ engine.gte = function(a, b){
...
+  if(b.length > 1) {
+    throw new Error("Expected singleton on right side of contains, got " + JSON.stringify(b));
+  }
+  return containsImpl(a,b);
+};
+
+engine.in = function(a, b){
+  if(a.length > 1) {
+    throw new Error("Expected singleton on right side of in, got " + JSON.stringify(b));
+  }
```
[the remark this answers is not in the document]
Comment (plynchnlm) [reply src/equality.js]: I think we could use util.assertAtMostOne, or something like that here.

File: src/equality.js (hunk 1 above)
Comment (niquola) [reply src/equality.js]: done

File: src/equality.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (niquola) [reply src/equality.js]: Yes, i did this to make message more reasonable

File: src/equality.js [hunk 3]
```diff
@@ -69,5 +69,26 @@ engine.gte = function(a, b){
...
 
+function containsImpl(a,b){
+  if(b.length == 0) { return true; }
+  for(var i = 0; i < a.length; i++){
+    if(deepEqual(a[i], b[0])) { return true; }
+  }
+  return false;
+}
+
+engine.contains = function(a, b){
```
[the remark this answers is not in the document]
Comment (niquola) [reply src/equality.js]: moved to collections module

File: src/equality.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (niquola) [reply src/equality.js]: Yes, i did this to make message more reasonable

File: src/collections.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+engine.contains = function(a, b){
+  if(b.length > 1) {
+    throw new Error("Expected singleton on right side of contains, got " + JSON.stringify(b));
+  }
+  return containsImpl(a,b);
+};
+
+engine.in = function(a, b){
```
Comment (xluanlhc) [inline src/collections.js]: Looks like the code has been moved(?), reposting the comment here:
When a is empty, the function should return empty per spec.

File: src/collections.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+var util = require("./utilities");
+var deepEqual = require('./deep-equal');
+
+var engine = {};
+
+
+// b is assumed to have one element and it tests whether b[0] is in a
+function containsImpl(a,b){
+  if(b.length == 0) { return true; }
```
Comment (plynchnlm) [inline src/collections.js]: Per Xioacheng's comment, maybe this should return [ ] ?  Well, I see you are always returning a boolean.  What can you return that will result in [ ] for the those cases?

File: src/collections.js (hunk 4 above)
Comment (niquola) [reply src/collections.js]: It's done automatically by infixInvoke function and nullable property

Comment (niquola) [reply src/collections.js]: there are tests for it

File: src/collections.js (hunk 5 above)
Comment (niquola) [reply src/collections.js]: see comment above

Comment (niquola) [reply src/collections.js]: https://github.com/lhncbc/fhirpath.js/blob/dea4597c4b29317d9dcc3f3120acf551ce83f858/test/cases/6.4_collections.yaml#L20

File: src/collections.js (hunk 4 above)
Comment (niquola) [reply src/collections.js]: https://github.com/lhncbc/fhirpath.js/blob/dea4597c4b29317d9dcc3f3120acf551ce83f858/test/cases/6.4_collections.yaml#L20

Comment (plynchnlm) [reply src/collections.js]: infixInvoke returns  [ ] if either parameter is null.  This won't work for "in" where there is a difference in return value depending on whether the left or the right parameter is null.
```
bin/fhirpath '1.combine(2) in {}' <(echo '{}')
```
should return [false], but it returns [ ].

**Comment by plynchnlm** [comment]
New problem:  "If the left operand (of "in") has multiple items, an exception is thrown."
```
$ bin/fhirpath '1.combine(2) in {}' <(echo '{}')
fhirpath(1.combine(2) in {}) =>
[
 false
]
```
Please check whether contains has similar issues.

**Pull request merged**